### ISSUE TITLE
MapParallelOrdered is not running all Tasks in parallel

### ISSUE THREAD

**Issue by lukestephenson** [monix/monix#841]
This was originally mentioned to me by @cwmyers who I work with.  I've since looked into it a little more and thought about how it could be fixed.

# Problem / Issue description

With the following simple Observable using `mapParallelUnordered`, the process takes just over a second to complete as the Tasks are executed in parallel.
```
package monix

import monix.eval.Task
import monix.execution.Scheduler
import monix.reactive.Observable

import scala.concurrent.Await
import scala.concurrent.duration.Duration

object Example2 {
  implicit val scheduler = Scheduler.fixedPool("MyTest", 20)

  def createStream = {
    val last =
    Observable.range(1, 5)
        .mapParallelUnordered(20)(waitForABit)
        .dump("Int = ")
        .runAsyncGetLast
    val maybeInt = Await.result(last, Duration.Inf)
    maybeInt
  }

  def waitForABit(i: Long): Task[Long] = {
    Task {
      Thread.sleep(1000)
      println(Thread.currentThread().getName + s" - $i")
      i
    }
  }

  def main(args: Array[String]): Unit = {
    createStream
  }
}
```

Note the ordering is not guaranteed (as expected).
```
MyTest-14 - 4
MyTest-11 - 1
MyTest-12 - 2
MyTest-13 - 3
0: Int =  --> 3
1: Int =  --> 4
2: Int =  --> 2
3: Int =  --> 1
4: Int =  completed
```

Now if the above program is changed from `mapParallelUnordered` to `mapParallelOrdered` the behaviour changes significantly (beyond just the ordered output):

1. Everything occurs on the `main` thread.
```
main - 1
0: Int =  --> 1
main - 2
1: Int =  --> 2
main - 3
2: Int =  --> 3
main - 4
3: Int =  --> 4
4: Int =  completed
```
2. Execution is sequential
This program takes just over 4 seconds to complete.

While the use of `Thread.sleep` isn't ideal and replacing with a `Task.delayExecution` would fix the behaviour in this example, I don't think the current implementation of `mapParallelOrdered` is ideal.

I think users of the Monix library aren't expecting a radically different result when switching between `mapParallelOrdered` and `mapParallelUnordered`

Also, if the async Task was CPU bound, the implementation would also fail to achieve parallelism.

As such I'm proposing the current implementation is changed.

# Potential issue
Currently the elements are `process`ed when the parallelism semaphore permit is acquired.
https://github.com/monix/monix/blob/adef8b31dd612e9504aab6559dc8cacb985a7491/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala#L170-L181
This is done on the same thread if the semaphore is acquired synchronously.

Compare this to where `process` is called for the `Unordered` implementation.
https://github.com/monix/monix/blob/adef8b31dd612e9504aab6559dc8cacb985a7491/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala#L170-L181
Here it's done in `ack.onComplete` for which the callback occurs on a different thread.

# Proposed fix
`process` should be called in `ack.onComplete` just like it is in the `Unordered` implementation so that the tasks are invoked on the Scheduler / ExecutionContext rather than the current thread.

Note that `process` currently is also responsible for managing a `queue` of the tasks so that they can be notified in order.  If `process` is called in `ack.onComplete`, the order can no longer be guaranteed at that point, so the items must be added to the `queue` earlier.  I suggest changing the queue to wrap the futures in `AtomicAny` so that the ordering can be set when it is guaranteed (and the AtomicAny references to be set when the Task is started).

I've started experimenting with this approach, but wanted to give people a chance to comment on the issue before I submit a PR.  I'm also keen to hear thoughts on how to test the change.

**Comment by Avasil** [comment]
Thanks a lot for the extensive report @lukestephenson . I will look into that and think about possible solutions. PR is of course very welcome, we can go from there.

The first possible solution that comes to my mind and might work is to force async boundary on executed tasks 

Not sure how to test it yet

**Comment by lukestephenson** [comment]
@Avasil Thanks for the suggestion to force an async boundary. I've given that a try as well as what I initially described.  Both approaches are currently failing on the same cancellation tests, which I still need to investigate.  I'll submit the PRs for comment now (with the broken tests).

**Comment by lukestephenson** [comment]
@Avasil The approach you suggested appears to be working (but failing on an unrelated test).

**Comment by Avasil** [comment]
@lukestephenson I've restarted CI and it's ok for both tests now. :)

I feel like the async boundary should be better if there aren't more corner cases. Wrapping everything in `Atomic` is probably expensive but I'm not sure without benchmark

**Comment by lukestephenson** [comment]
Hi @Avasil .  I'm happy with the Async boundary approach.

The only corner case I wasn't sure about is if there can be multiple calls to `onNext` waiting for an async semaphore permit at the same time.  But the change doesn't make that any worse if it is a problem.

**Comment by Avasil** [comment]
> The only corner case I wasn't sure about is if there can be multiple calls to `onNext` waiting for an async semaphore permit at the same time. But the change doesn't make that any worse if it is a problem.

IIRC there is an internal `Queue` in `Semaphore` implementation so the order should be preserved

**Comment by lukestephenson** [comment]
What I was concerned about is https://github.com/monix/monix/blob/0097879cac2b5355bcc4b9c33046bf92ec111c42/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala#L170-L181:
```
val ack: Future[Ack] = permit.value match {
  case None =>
    permit.flatMap { _ =>
      composite -= permit
      process(elem)
      Continue
    }
  case Some(_) =>
    composite -= permit
    process(elem)
    Continue
}
```

When the permit is obtained synchronously (ie `case Some(_)`), then ordering is guaranteed.
When it is obtained asynchronously (ie `case None`), then the logic is performed in a `future.flatMap` callback which will be run potentially on a separate Thread under the control of the ExecutionContext.  If there are multiple permits released asynchronously at once, then I don't think there is a guarantee on the ordering that those callbacks will occur.  But I'm not sure if it's even possible to get into a state where multiple permits are released asynchronously at once.  Either way, the change I've implemented wouldn't make that worse.

Hope that makes sense.

**Comment by Avasil** [comment]
Ah, I see what you mean. I'll check if I can reproduce it and fix if it turns out to be a problem. Thanks! 

And regarding this issue, I guess now we just have to figure out a way to test the current fix

**Comment by lukestephenson** [comment]
This was fixed by #846
